cpp/src/arrow/json/chunker.cc

```diff
@@ -187,88 +118,60 @@ static size_t ConsumeWholeObject(Stream&& stream) {
   }
 }
 
-// A chunker implementation that assumes JSON objects can contain raw newlines,
-// and uses actual JSON parsing to delimit chunks.
-class ParsingChunker : public Chunker {
+namespace {
+
+// A delimiter implementation that assumes JSON objects can contain raw newlines,
+// and uses actual JSON parsing to delimit them.
+class ParsingDelimiter : public Delimiter {
  public:
-  Status Process(std::shared_ptr<Buffer> block, std::shared_ptr<Buffer>* whole,
-                 std::shared_ptr<Buffer>* partial) override {
-    if (block->size() == 0) {
-      *whole = SliceBuffer(block, 0, 0);
-      *partial = block;
-      return Status::OK();
+  Status FindFirst(string_view partial, string_view block, int64_t* out_pos) override {
+    // XXX should we bubble up JSON parse errors instead of ignoring them?
```

Comment:
I think it's fine to wait; in the (unusual) case of a parse error we'll just fail a little later by deferring those checks to the thread pool